File: packages/typescript/src/platforms/threads.ts
/**
 * Threads adapter.
 *
 * Notes:
 *   - Login uses Instagram credentials (Threads is an Instagram product).
 *     The adapter clicks "Continue with Instagram" and reuses IG cookies if
 *     the IG context is already authenticated; else falls back to entering
 *     IG credentials directly on threads.net.
 *   - "Follow" maps to following on Threads (which mirrors IG follow).
 */

import { BasePlatform } from "../core/BasePlatform.js";
import { resolveMfa } from "../core/mfa.js";
import type {
  CommentInput,
  CommentResult,
  Credentials,
  FollowInput,
  FollowResult,
  LikeInput,
  LikeResult,
  PostInput,
  PostResult,
} from "../core/types.js";

export class ThreadsAdapter extends BasePlatform {
  protected loginUrl(): string {
    return "https://www.threads.net/login";
  }

  protected async performLogin(creds: Credentials): Promise<void> {
    const sh = await this.getStagehand();
    // Click "Continue with Instagram" if available.
    await sh.page.act("Click 'Continue with Instagram' if visible").catch(() => undefined);
    await sh.page.waitForLoadState("networkidle").catch(() => {});

    // If IG cookie carryover already logged us in, exit.
    if (await this.isLoggedIn().catch(() => false)) return;

    // Else fall through: type IG credentials on Threads' login page.
    await sh.page.act(`Type "${creds.username}" into the username field`);
    await sh.page.act(`Type "${creds.password}" into the password field`);
    await sh.page.act("Click the Log in button");
    await sh.page.waitForLoadState("networkidle").catch(() => {});

    const mfa = await sh.page.extract({
      instruction: "Is Threads / Instagram asking for a 2FA code? Return mfa=true|false.",
      schema: { mfa: "boolean" },
    }).catch(() => ({ mfa: false }));
    if ((mfa as { mfa: boolean }).mfa) {
      const code = await resolveMfa(
        { description: "Threads 2FA (Instagram-backed)", channel: "unknown" },
        creds,
        this.onMfaChallenge,
      );
      await sh.page.act(`Type "${code}" into the verification code field`);
      await sh.page.act("Click Confirm");
      await sh.page.waitForLoadState("networkidle").catch(() => {});
    }
  }

  async isLoggedIn(): Promise<boolean> {
    const sh = await this.getStagehand();
    if (!sh.page.url().includes("threads.net")) {
      await sh.page.goto("https://www.threads.net/").catch(() => {});
    }
    const r = await sh.page.extract({
      instruction:
        "Is the user logged into Threads? Look for a profile avatar in the bottom nav, a 'New thread' button, or the home feed. Return loggedIn=true|false.",
      schema: { loggedIn: "boolean" },
    }).catch(() => ({ loggedIn: false }));
    return Boolean((r as { loggedIn: boolean }).loggedIn);
  }

  async post(input: PostInput): Promise<PostResult> {
    const sh = await this.getStagehand();
    await sh.page.goto("https://www.threads.net/");
    await sh.page.act("Click the 'New thread' button (pencil icon) in the left sidebar or bottom nav");

    if (input.media?.length) {
      const chooserPromise = sh.page.waitForEvent("filechooser");
      await sh.page.act("Click the attach (paperclip / photo) button in the new-thread composer");
      const chooser = await chooserPromise;
      await chooser.setFiles(input.media);
    }

    if (input.text) {
      const body = this.composeCaption(input);
      await sh.page.act(`Type the following into the thread body: ${JSON.stringify(body)}`);
    }

    await sh.page.act("Click the 'Post' button to publish the thread");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now() };
  }

  private composeCaption(input: PostInput): string {
    let body = input.text ?? "";
    if (input.mentions?.length) {
      body += " " + input.mentions.map((m) => `@${m.replace(/^@/, "")}`).join(" ");
    }
    if (input.hashtags?.length) {
      body += " " + input.hashtags.map((h) => `#${h.replace(/^#/, "")}`).join(" ");
    }
    return body.trim();
  }

  async comment(input: CommentInput): Promise<CommentResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the 'Reply' input under the thread");
    await sh.page.act(`Type the reply: ${JSON.stringify(input.text)}`);
    await sh.page.act("Click the 'Post' button to submit the reply");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url };
  }

  async like(input: LikeInput): Promise<LikeResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the heart icon under the thread to like it");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url, liked: true };
  }

  async unlike(input: LikeInput): Promise<LikeResult> {
    const sh = await this.getStagehand();
    await sh.page.goto(input.url);
    await sh.page.act("Click the (already-filled red) heart icon under the thread to unlike it");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), url: input.url, liked: false };
  }

  async follow(input: FollowInput): Promise<FollowResult> {
    const sh = await this.getStagehand();
    const handle = input.username.replace(/^@/, "");
    await sh.page.goto(`https://www.threads.net/@${handle}`);
    await sh.page.act("Click the 'Follow' button on the profile header");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), following: true };
  }

  async unfollow(input: FollowInput): Promise<FollowResult> {
    const sh = await this.getStagehand();
    const handle = input.username.replace(/^@/, "");
    await sh.page.goto(`https://www.threads.net/@${handle}`);
    await sh.page.act("Click the 'Following' button on the profile header, then click 'Unfollow' in the dialog");
    await sh.page.waitForLoadState("networkidle").catch(() => {});
    return { success: true, at: this.now(), following: false };
  }
}

File: packages/python/stagehand_social/platforms/threads.py
"""Threads adapter — Python mirror of platforms/threads.ts."""

from __future__ import annotations

import json

from stagehand_social.core.base_platform import BasePlatform
from stagehand_social.core.mfa import resolve_mfa
from stagehand_social.core.types import (
    CommentInput,
    CommentResult,
    Credentials,
    FollowInput,
    FollowResult,
    LikeInput,
    LikeResult,
    MfaChallenge,
    PostInput,
    PostResult,
)


class ThreadsAdapter(BasePlatform):
    def login_url(self) -> str:
        return "https://www.threads.net/login"

    async def perform_login(self, creds: Credentials) -> None:
        sh = await self.get_stagehand()
        try:
            await sh.page.act("Click 'Continue with Instagram' if visible")
        except Exception:
            pass
        try:
            await sh.page.wait_for_load_state("networkidle")
        except Exception:
            pass

        # If IG cookie carryover already logged us in, exit.
        try:
            already = await self.is_logged_in()
        except Exception:
            already = False
        if already:
            return

        await sh.page.act(f'Type "{creds.username}" into the username field')
        await sh.page.act(f'Type "{creds.password}" into the password field')
        await sh.page.act("Click the Log in button")
        try:
            await sh.page.wait_for_load_state("networkidle")
        except Exception:
            pass

        try:
            mfa = await sh.page.extract(
                instruction="Is Threads / Instagram asking for a 2FA code? Return mfa=true|false.",
                schema={"mfa": "boolean"},
            )
        except Exception:
            mfa = {"mfa": False}
        if isinstance(mfa, dict) and mfa.get("mfa"):
            code = await resolve_mfa(
                MfaChallenge(description="Threads 2FA (Instagram-backed)", channel="unknown"),
                creds,
                self.on_mfa_challenge,
            )
            await sh.page.act(f'Type "{code}" into the verification code field')
            await sh.page.act("Click Confirm")

    async def is_logged_in(self) -> bool:
        sh = await self.get_stagehand()
        if "threads.net" not in sh.page.url:
            try:
                await sh.page.goto("https://www.threads.net/")
            except Exception:
                pass
        try:
            r = await sh.page.extract(
                instruction=(
                    "Is the user logged into Threads? Look for a profile avatar, 'New thread' button, "
                    "or the home feed. Return loggedIn=true|false."
                ),
                schema={"loggedIn": "boolean"},
            )
            return bool(isinstance(r, dict) and r.get("loggedIn"))
        except Exception:
            return False

    async def post(self, input: PostInput) -> PostResult:
        sh = await self.get_stagehand()
        await sh.page.goto("https://www.threads.net/")
        await sh.page.act("Click the 'New thread' button (pencil icon) in the left sidebar or bottom nav")

        if input.media:
            async with sh.page.expect_file_chooser() as fc_info:
                await sh.page.act("Click the attach (paperclip / photo) button in the new-thread composer")
            chooser = await fc_info.value
            await chooser.set_files(input.media)

        if input.text:
            body = self._compose_caption(input)
            await sh.page.act(f"Type the following into the thread body: {json.dumps(body)}")

        await sh.page.act("Click the 'Post' button to publish the thread")
        return PostResult(success=True, at=self._now())

    def _compose_caption(self, input: PostInput) -> str:
        body = input.text or ""
        if input.mentions:
            body += " " + " ".join(f"@{m.lstrip('@')}" for m in input.mentions)
        if input.hashtags:
            body += " " + " ".join(f"#{h.lstrip('#')}" for h in input.hashtags)
        return body.strip()

    async def comment(self, input: CommentInput) -> CommentResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the 'Reply' input under the thread")
        await sh.page.act(f"Type the reply: {json.dumps(input.text)}")
        await sh.page.act("Click the 'Post' button to submit the reply")
        return CommentResult(success=True, at=self._now(), url=input.url)

    async def like(self, input: LikeInput) -> LikeResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the heart icon under the thread to like it")
        return LikeResult(success=True, at=self._now(), url=input.url, liked=True)

    async def unlike(self, input: LikeInput) -> LikeResult:
        sh = await self.get_stagehand()
        await sh.page.goto(input.url)
        await sh.page.act("Click the (already-filled red) heart icon under the thread to unlike it")
        return LikeResult(success=True, at=self._now(), url=input.url, liked=False)

    async def follow(self, input: FollowInput) -> FollowResult:
        sh = await self.get_stagehand()
        handle = input.username.lstrip("@")
        await sh.page.goto(f"https://www.threads.net/@{handle}")
        await sh.page.act("Click the 'Follow' button on the profile header")
        return FollowResult(success=True, at=self._now(), following=True)

    async def unfollow(self, input: FollowInput) -> FollowResult:
        sh = await self.get_stagehand()
        handle = input.username.lstrip("@")
        await sh.page.goto(f"https://www.threads.net/@{handle}")
        await sh.page.act("Click the 'Following' button on the profile header, then click 'Unfollow' in the dialog")
        return FollowResult(success=True, at=self._now(), following=False)
